File: create_portfolio.py
#!/usr/bin/env python3
"""
Script to create a Portfolio in Monday.com using GraphQL API.
Creates a portfolio named "Duetto" (Salesforce account name) in workspace ID 13800719.

Usage:
    MONDAY_API_KEY="your-api-key" python3 create_portfolio.py
"""

import os
import sys
import json
import requests

# Configuration
WORKSPACE_ID = "13800719"
PORTFOLIO_NAME = "Duetto"  # Salesforce account name
API_URL = "https://api.monday.com/v2"


def create_portfolio(api_key: str, workspace_id: str, name: str) -> dict:
    """
    Create a portfolio (folder) in Monday.com using GraphQL API.
    
    Args:
        api_key: Monday.com API key
        workspace_id: ID of the workspace to create the portfolio in
        name: Name of the portfolio to create
        
    Returns:
        dict: Response data from the API
    """
    headers = {
        "Authorization": api_key,
        "Content-Type": "application/json",
        "API-Version": "2024-10"
    }
    
    # GraphQL mutation to create a portfolio in Monday.com
    # Note: folder_kind: portfolio is required to create a portfolio (not just a folder)
    mutation = """
    mutation CreatePortfolio($workspace_id: ID!, $name: String!) {
        create_folder(workspace_id: $workspace_id, name: $name, folder_kind: portfolio) {
            id
            name
        }
    }
    """
    
    variables = {
        "workspace_id": workspace_id,
        "name": name
    }
    
    payload = {
        "query": mutation,
        "variables": variables
    }
    
    response = requests.post(API_URL, headers=headers, json=payload)
    response.raise_for_status()
    
    return response.json()


def main():
    # Get API key from environment variable
    api_key = os.environ.get("MONDAY_API_KEY")
    
    if not api_key:
        print("Error: MONDAY_API_KEY environment variable is required", file=sys.stderr)
        print(f"Usage: MONDAY_API_KEY=\"your-api-key\" python3 {sys.argv[0]}", file=sys.stderr)
        sys.exit(1)
    
    print(f"Creating Portfolio '{PORTFOLIO_NAME}' in workspace ID {WORKSPACE_ID}...")
    
    try:
        result = create_portfolio(api_key, WORKSPACE_ID, PORTFOLIO_NAME)
        
        if "errors" in result:
            print(f"\nGraphQL Errors:")
            print(json.dumps(result["errors"], indent=2))
            sys.exit(1)
        
        if "data" in result and result["data"].get("create_folder"):
            folder = result["data"]["create_folder"]
            print(f"\nSuccessfully created Portfolio!")
            print(f"Portfolio ID: {folder['id']}")
            print(f"Portfolio Name: {folder['name']}")
        else:
            print(f"\nUnexpected response format:")
            print(json.dumps(result, indent=2))
            sys.exit(1)
            
    except requests.exceptions.RequestException as e:
        print(f"\nHTTP Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: create_portfolio.js
#!/usr/bin/env node

/**
 * Script to create a Portfolio in Monday.com using GraphQL API.
 * Creates a portfolio named "Duetto" (Salesforce account name) in workspace ID 13800719.
 * 
 * Usage:
 *   MONDAY_API_KEY=your-api-key node create_portfolio.js
 */

const https = require('https');

// Configuration
const MONDAY_API_URL = 'api.monday.com';
const PORTFOLIO_NAME = 'Duetto';  // Salesforce account name
const WORKSPACE_ID = '13800719';

/**
 * Create a portfolio in Monday.com using GraphQL API.
 * Note: folder_kind: portfolio is required to create a portfolio (not just a folder)
 * 
 * @param {string} apiKey - Monday.com API key
 * @param {string} portfolioName - Name for the portfolio
 * @param {string} workspaceId - ID of the workspace to create the portfolio in
 * @returns {Promise<object>} API response
 */
async function createPortfolio(apiKey, portfolioName, workspaceId) {
    // GraphQL mutation to create a portfolio
    const graphqlQuery = `
        mutation CreatePortfolio($workspace_id: ID!, $name: String!) {
            create_folder(workspace_id: $workspace_id, name: $name, folder_kind: portfolio) {
                id
                name
            }
        }
    `;

    const payload = JSON.stringify({
        query: graphqlQuery,
        variables: {
            workspace_id: workspaceId,
            name: portfolioName
        }
    });

    const options = {
        hostname: MONDAY_API_URL,
        port: 443,
        path: '/v2',
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Authorization': apiKey,
            'API-Version': '2024-10',
            'Content-Length': Buffer.byteLength(payload)
        }
    };

    return new Promise((resolve, reject) => {
        const req = https.request(options, (res) => {
            let data = '';

            res.on('data', (chunk) => {
                data += chunk;
            });

            res.on('end', () => {
                try {
                    const response = JSON.parse(data);
                    resolve(response);
                } catch (e) {
                    reject(new Error(`Failed to parse response: ${data}`));
                }
            });
        });

        req.on('error', (error) => {
            reject(error);
        });

        req.write(payload);
        req.end();
    });
}

async function main() {
    // Get API key from environment variable
    const apiKey = process.env.MONDAY_API_KEY;

    if (!apiKey) {
        console.error('Error: MONDAY_API_KEY environment variable is required');
        console.error('Usage: MONDAY_API_KEY=your-api-key node create_portfolio.js');
        process.exit(1);
    }

    console.log(`Creating Portfolio '${PORTFOLIO_NAME}' in workspace ID ${WORKSPACE_ID}...`);

    try {
        const response = await createPortfolio(apiKey, PORTFOLIO_NAME, WORKSPACE_ID);

        // Check for errors in the response
        if (response.errors) {
            console.error('\nGraphQL Errors:');
            response.errors.forEach(error => {
                console.error(`  - ${error.message || JSON.stringify(error)}`);
            });
            process.exit(1);
        }

        // Extract portfolio data
        const portfolioData = response.data?.create_folder;

        if (portfolioData) {
            console.log('\nSuccessfully created Portfolio!');
            console.log(`Portfolio ID: ${portfolioData.id}`);
            console.log(`Portfolio Name: ${portfolioData.name}`);
        } else {
            console.error('\nUnexpected response format:');
            console.log(JSON.stringify(response, null, 2));
            process.exit(1);
        }
    } catch (error) {
        console.error(`\nError: ${error.message}`);
        process.exit(1);
    }
}

main();
